### ISSUE TITLE
SQL cache is unusably slow

### ISSUE THREAD

**Issue by edenman** [apollographql/apollo-kotlin#5834]
### Version

4.0.0-beta.4

### Summary

- New project, interacting with a old and very gnarly graphql schema. 
- Response is ~200kb
- We're observing the cache with a `CacheOnly` policy and `watch`
- When we make a network call, we're using `NetworkOnly` and then manually manipulating the response before writing it to the cache
- When we write a response to the cache, it takes *14* seconds for the `watch` to fire.  If I remove the `SqlNormalizedCacheFactory` that drops down to 3 seconds.

### Steps to reproduce the behavior

```
        ApolloClient.Builder().apply {
            interceptors(apolloInterceptors.toList())
            networkTransport(
                HttpNetworkTransport.Builder()
                    .httpEngine(DefaultHttpEngine(okHttpClient))
                    .httpRequestComposer(RequestComposer)
                    .build(),
            )
            normalizedCache(
                MemoryCacheFactory(10 * 1024 * 1024)
                    .chain(SqlNormalizedCacheFactory("$userId-apollo.db"))
            )
            writeToCacheAsynchronously(true)
            autoPersistedQueries()
        }.build()
```

**Comment by BoD** [comment]
Hi! Thanks for reporting.

Can you tell us a bit more about the response manipulation you're doing, and how it is done? Are you using `doNotStore` so that the data is not written twice to the cache? Is the writing call itself taking a while, or is it only the watch being slow to be triggered (or both).

Also in general if you could share your (or part of your) schema, query and json payload, that would help a lot to have a look and try to understand what is happening (share privately to benoit.lubek@apollographql.com if that's preferable).

**Comment by edenman** [comment]
- Correct, we're doing `doNotStore(true)` on the network query.  
- The response is a list of heterogeneous items
- The response includes some special caching instructions in the body, so we're manually processing those and then merging the existing items with the new ones, and then inserting a new FooResponse into the cache
- I'll send you the json payload via email

**Comment by BoD** [comment]
Thanks a lot for having shared all the details. I could reproduce writing to the cache taking an unusually long time, due to a transaction per insert being used, instead of just one transaction for all of them, when the memory cache is used with chaining. This is fixed [here](https://github.com/apollographql/apollo-kotlin/pull/5840) and will be in the next v4 release (planned in the next few days).

**Comment by BoD** [comment]
Released in 3.8.4
